File: breakout_zig/src/platform.zig
const rl = @import("raylib.zig");
const collision = @import("collision.zig");
const Box = @import("box.zig").Box;
const Scene = @import("scene.zig").Scene;

const Vector2 = rl.Vector2;
const Color = rl.Color;
const IsKeyDown = rl.IsKeyDown;
const KEY_LEFT = rl.KEY_LEFT;
const KEY_RIGHT = rl.KEY_RIGHT;

const Collider = collision.Collider;
const Collision = collision.Collision;

pub const Platform = struct {
    const Self = @This();

    box: Box,
    speed: f32,

    pub fn new(center: Vector2, size: Vector2, color: Color, speed: f32) Self {
        return Self{
            .box = Box.new(
                center,
                size,
                color,
            ),
            .speed = speed,
        };
    }

    pub fn draw(self: *const Self) void {
        self.box.draw();
    }

    pub fn collides(self: *const Self, collider: *const Collider) ?Collision {
        return self.box.collides(collider);
    }

    pub fn update(self: *Self, scene: *const Scene, dt: f32) void {
        if (IsKeyDown(KEY_LEFT)) {
            self.box.collider.rect.move(Vector2{ .x = -self.speed * dt, .y = 0 });
        }
        if (IsKeyDown(KEY_RIGHT)) {
            self.box.collider.rect.move(Vector2{ .x = self.speed * dt, .y = 0 });
        }

        const col = scene.collides(&self.box.collider);
        if (col) |c| {
            if (0.0 < c.normal.x) {
                self.box.collider.rect.set_pos(Vector2{ .x = c.pos.x + self.box.collider.rect.width() / 2.0, .y = self.box.collider.rect.pos().y });
            } else {
                self.box.collider.rect.set_pos(Vector2{ .x = c.pos.x - self.box.collider.rect.width() / 2.0, .y = self.box.collider.rect.pos().y });
            }
        }
    }
};

File: breakout_go/src/platform.go
package src

import (
	rl "github.com/gen2brain/raylib-go/raylib"
)

type Platform struct {
	rectangle Rect
	color     rl.Color
	speed     float32
}

func NewPlatform(center rl.Vector2, width float32, height float32, color rl.Color, speed float32) Platform {
	var rectangle = rect_from_center(center, width, height)
	return Platform{
		rectangle: rectangle,
		color:     color,
		speed:     speed,
	}
}

func (self *Platform) Draw() {
	rl.DrawRectangleRec(self.rectangle.raylib_rect(), self.color)
}

func (self *Platform) Update(scene *SceneBorder, dt float32) {
	if rl.IsKeyDown(rl.KeyA) {
		self.rectangle.X -= self.speed * dt
	}
	if rl.IsKeyDown(rl.KeyD) {
		self.rectangle.X += self.speed * dt
	}
	var collision_opt = scene.collides(self)
	if collision_opt.is_some() {
		var collision = collision_opt.unwrap()
		if 0.0 <= collision.normal.X {
			self.rectangle.X = collision.pos.X
		} else {
			self.rectangle.X = collision.pos.X - self.rectangle.Width
		}
	}
}

// impl Collider
func (self *Platform) rect() Option[*Rect] {
	return option_some(&self.rectangle)
}

func (self *Platform) collides(other Collider) Option[Collision] {
	return self.rectangle.collides(other)
}
